batch/batch/front_end/front_end.py

```diff
@@ -1496,6 +1534,81 @@ async def insert(tx):
     return web.HTTPFound(deploy_config.external_url('batch', '/billing_projects'))
 
 
+async def _close_billing_project(db, billing_project):
+    @transaction(db)
+    async def close_project(tx):
+        row = await tx.execute_and_fetchone(
+            '''
+SELECT name, `status`,
+  (SELECT 1 FROM batches WHERE
+    time_completed IS NULL AND
+    billing_project = billing_projects.name
+    LIMIT 1) AS batch
+FROM billing_projects WHERE name = %s and `status` != 'deleted' FOR UPDATE;
```

Comment:
```sql
select name, `status`, batches.id as batch_id
from billing_projects
left join batches
on billing_projects.name = batches.billing_project
where name = %s and `status` != 'deleted' and batches.time_completed is NULL
limit 1
FOR UPDATE
```
I suppose you might be intentionally not locking the batches table? I guess this is safe because time_completed should only transition from NULL to not NULL and if it happens to transition from NULL to not NULL while you're executing this transaction, that doesn't change the validity of this transaction.

Reply:
Yeah, that's true. I wasn't intentionally not locking it; it would still be good to lock it to prevent batch creation during this transaction, in fact, but again, I think I'd like to say "shared lock for this table and exclusive lock for this one, please" inside the join.